drivers/usb_c/tcpc/ucpd_numaker.c

```diff
@@ -0,0 +1,2163 @@
+/*
+ * Copyright (c) 2024 Nuvoton Technology Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT nuvoton_numaker_tcpc
+
+#include <zephyr/kernel.h>
+#include <zephyr/drivers/usb_c/usbc_tcpc.h>
+#include <zephyr/drivers/usb_c/usbc_ppc.h>
+#include <zephyr/drivers/clock_control.h>
+#include <zephyr/drivers/clock_control/clock_control_numaker.h>
+#include <zephyr/drivers/reset.h>
+#include <zephyr/drivers/pinctrl.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/drivers/adc.h>
+#include <zephyr/sys/byteorder.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(tcpc_numaker, CONFIG_USBC_LOG_LEVEL);
+
+#include <soc.h>
+#include <NuMicro.h>
+
+/* Implementation notes on NuMaker TCPC/PPC/VBUS
+ *
+ * 1. UTCPD, interfacing to external circuit on VBUS/VCONN voltage measurement,
+ *    VBUS/VCONN overcurrent protection, VBUS overvoltage protection, etc.,
+ *    can implement all functions defined in TCPC, PPC, and VBUS. For this,
+ *    TCPC is implemented in UTCPD majorly; PPC and VBUS rely on TCPC for
+ *    their implementation.
+ * 2. For VBUS/VCONN voltage measurement, UTCPD is updated periodically
+ *    by Timer-trigger EADC. To implement this interconnection, TCPC node_id
+ *    will cover UTCPD, EADC, and Timer H/W characteristics of registers,
+ *    interrupts, resets, and clocks.
+ *    NOTE: EADC and Timer interrupts needn't enable for Timer-triggered EADC.
+ *          In BSP sample, they are enabled just for development/debug purpose.
+ * 3. About VCONN per PCB
+ *    (1) Support only VCONN source, no VCONN sink (like Plug Cable)
+ *    (2) Separate pins for VCONN enable on CC1/CC2 (VCNEN1/VCNEN2)
+ *    (3) Single pin for VCONN discharge (DISCHG)
+ * 4. VBUS discharge precedence
+ *    (1) GPIO
+ *    (2) UTCPD
+ * 5. VCONN discharge precedence
+ *    (1) DPM-supplied callback
+ *    (2) GPIO
+ *    (3) UTCPD
+ */
+
+/* Invalid or missing value */
+#define NUMAKER_INVALID_VALUE ((uint32_t)-1)
+
+/* UTCPD VBUS threshold default in mV
+ *
+ * These are default values of UTCPD VBUS threshold registers. They need
+ * to be reconfigured by taking the following factors into consideration:
+ * 1. Analog Vref
+ * 2. UTCPD VBVOL.VBSCALE
+ */
+#define NUMAKER_UTCPD_VBUS_THRESHOLD_OVERVOLTAGE          25000
+#define NUMAKER_UTCPD_VBUS_THRESHOLD_VSAFE5V              5000
+#define NUMAKER_UTCPD_VBUS_THRESHOLD_VSAFE0V              0
+#define NUMAKER_UTCPD_VBUS_THRESHOLD_STOP_FORCE_DISCHARGE 800
+#define NUMAKER_UTCPD_VBUS_THRESHOLD_SINK_DISCONNECT      3500
```

Comment:
should this have a unit in the name? `NUMAKER_UTCPD_VBUS_THRESHOLD_OVERVOLTAGE_MV` etc

Reply:
Fixed